### PR TITLE
feat(query): support databend-local

### LINKED ISSUE #9281: Feature: databend-local
**Summary**

Like [clickhouse-local](https://clickhouse.com/docs/en/operations/utilities/clickhouse-local/), databend-local can perform fast processing on local files without launching a databend cluster.

### PR DESCRIPTION
**Pull request by eastfisher**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

This PR makes a demo databend-local, for directly query local parquet files without launching a databend cluster.

Closes #9281

Current result:

```
> export CONFIG_FILE=tests/local/config/databend-local.toml
> cargo run --bin=databend-local -- --sql="SELECT * FROM tbl1" --table=tbl1=/path/to/databend/docs/public/data/books.parquet

sql: SELECT * FROM read_parquet('/path/to/databend/docs/public/data/books.parquet')
+--------------------------------+-----------------------+--------+
| title                          | author                | date   |
+--------------------------------+-----------------------+--------+
| "Transaction Processing"       | "Jim Gray"            | "1992" |
+--------------------------------+-----------------------+--------+
| "Readings in Database Systems" | "Michael Stonebraker" | "2004" |
+--------------------------------+-----------------------+--------+
| "Transaction Processing"       | "Jim Gray"            | "1992" |
+--------------------------------+-----------------------+--------+
| "Readings in Database Systems" | "Michael Stonebraker" | "2004" |
+--------------------------------+-----------------------+--------+
4 rows in set (0.0163 sec)
```

### CODE REVIEW

**Comment by eastfisher** [comment]
I reuse the http query service for local query, but it seems not a good idea, some components are unused.

File: src/binaries/Cargo.toml [hunk 1]
```diff
@@ -52,6 +52,7 @@ clap = { workspace = true }
 limits-rs = "0.1.0"
 openraft = { workspace = true }
 poem = { version = "1", features = ["rustls", "multipart", "compression"] }
+prettytable-rs = "^0.9"
```
Comment (PsiACE) [inline src/binaries/Cargo.toml:55]: comfy-table

File: src/binaries/local/main.rs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    let conf: Config = Config::load_for_test()?;
+
+    set_panic_hook();
+    MetaEmbedded::init_global_meta_store(conf.meta.embedded_dir.clone()).await?;
+
+    // Make sure global services have been inited.
+    GlobalServices::init(conf.clone()).await?;
+
+    let session = SessionManager::instance().create_session(SessionType::Local).await?;
+    let httpctx = HttpQueryContext::new(session.clone());
```
Comment (sundy-li) [inline src/binaries/local/main.rs:64]: Yes, you should query directly through:

```
  let mut planner = Planner::new(context.clone());
  let (plan, _, fmt) = planner.plan_sql(&sql).await.map_err(BadRequest)?;
  let format = get_format_with_default(fmt, default_format)?;

  context.attach_query_str(plan.to_string(), &sql);
  let interpreter = InterpreterFactory::get(context.clone(), &plan)
      .await
      .map_err(BadRequest)?;
  execute(context, interpreter, plan.schema(), format, params, None)

```

Maybe you don't need to pretty the block by yourself, you can pass the format into SQL, eg:

```
select * from tb1 format NDJSON
```

Comment (eastfisher) [reply src/binaries/local/main.rs:71]: Should we reuse the `clickhouse_handler` or write a new `local_handler`?

Comment (sundy-li) [reply src/binaries/local/main.rs:71]: I do not think we need any handlers. Just follow:

` SQL --> Parser --> Plan --> Interpreter --> execute --> blocks`

Comment (eastfisher) [reply src/binaries/local/main.rs]: Fixed, PTAL~

File: src/binaries/local/main.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    let first_block = match stream.next().await {
+        Some(block) => match block {
+            Ok(block) => block,
+            Err(err) => return Err(err),
+        },
+        None => return Err(ErrorCode::Internal("no data block return")),
+    };
+    let mut output_fmt = FileFormatOptionsExt::get_output_format_from_options(
+        first_block.schema().clone(),
+        FileFormatOptions::default(),
```
Comment (eastfisher) [inline src/binaries/local/main.rs:77]: Now the default CSV format is used.

**Comment by sundy-li** [comment]
can we make it a `subcomand` of databend-query? thus we don't need another binary file to release.

```
ln -s databend-query databend-local

## this will parse args in the path of local
./databend-local   
```

In the future, we may have `databend-cli` tools.

**Comment by eastfisher** [comment]
Now `local` is converted to query subcommand:

> databend-query --cmd local --sql "SELECT 'hello'"

PTAL~

**Comment by Xuanwo** [comment]
> databend-query --cmd local --sql "SELECT 'hello'"

Can we read from stdin so that we can `echo "SELECT 'hello'" | databend-query`?

**Comment by eastfisher** [comment]
> > databend-query --cmd local --sql "SELECT 'hello'"
> 
> Can we read from stdin so that we can `echo "SELECT 'hello'" | databend-query`?

Of course we can, but databend-local may have other args, the sql is one of the args.

**Comment by eastfisher** [comment]
https://clickhouse.com/docs/en/operations/utilities/clickhouse-local/

```bash
$ echo -e "1,2\n3,4" | clickhouse-local --structure "a Int64, b Int64" \
    --input-format "CSV" --query "SELECT * FROM table"
Read 2 rows, 32.00 B in 0.000 sec., 5182 rows/sec., 80.97 KiB/sec.
1   2
3   4
```

In clickhouse-local, The data can be read from stdin and the sql is cmd arg.

**Review by sundy-li — APPROVED** [review]

File: src/binaries/Cargo.toml (hunk 1 above)
Comment (eastfisher) [reply src/binaries/Cargo.toml]: Fixed~

**Comment by eastfisher** [comment]
Using comfy-table to format result and now it looks like this:

```
exec local query: SELECT * FROM tbl1 FORMAT XML
+------------------------------+---------------------+------+
| title                        | author              | date |
+------------------------------+---------------------+------+
| Transaction Processing       | Jim Gray            | 1992 |
| Readings in Database Systems | Michael Stonebraker | 2004 |
| Transaction Processing       | Jim Gray            | 1992 |
| Readings in Database Systems | Michael Stonebraker | 2004 |
+------------------------------+---------------------+------+
4 rows in set. Query took 0.015 seconds.
```

**Review by Xuanwo — APPROVED** [review]

**Comment by BohuTANG** [comment]
👍

**Pull request merged**